File: Python/Homework3/DLList.py
import random
from Node import Node


class DLList:
    def __init__(self):
        self.start = None
        self.end = None

    def print_dllist(self):
        temp = self.start
        s = ""
        while temp != None:
            s += str(temp.x) + " <=> "
            temp = temp.nextNode
        print(s[0:-5])

    def push_front(self, x: int):
        t = Node(x)
        if self.start == None:
            self.start = t
            self.end = t
            return
        t.nextNode = self.start
        self.start.prevNode = t
        self.start = t

    def insert(self, v: Node, x: int):
        t = Node(x)
        next_el = v.nextNode
        if next_el == None:
            t.nextNode = None
            v.nextNode = t
            t.prevNode = v
            self.end = t
            return
        next_el.prevNode = t
        t.nextNode = next_el
        v.nextNode = t
        t.prevNode = v

    def push_back(self, x: int):
        if self.end == None:
            t = Node(x)
            self.start = t
            self.end = t
            return
        DLList.insert(self, self.end, x)

    def delete(self, v: Node):
        prev = v.prevNode
        nextt = v.nextNode
        if prev != None:
            prev.nextNode = nextt
        if nextt != None:
            nextt.prevNode = prev
        v.nextNode = None
        v.prevNode = None
        if v == self.start:
            self.start = nextt
        if v == self.end:
            self.end = prev

    def pop_front(self):
        DLList.delete(self, self.start)

    def pop_back(self):
        DLList.delete(self, self.end)

    def revert(self):
        if (self.start == None):
            return
        a = self.start
        b = self.start.nextNode
        a.nextNode = None
        if b == None:
            return
        c = b.nextNode
        while b != None:
            b.nextNode = a
            a.prevNode = b
            a = b
            b = c
            if c != None:
                c = c.nextNode
        a.prevNode = None
        t = self.start
        self.start = self.end
        self.end = t

    def gen_list(self, count: int):
        a = 0
        b = 100

        for _ in range(count):
            value = random.randint(a, b)
            DLList.push_back(self, value)

File: Java/src/Homework3/DLList.java
package Homework3;

public class DLList {
    Node start;
    Node end;

    public DLList() {
        start = null;
        end = null;
    }

    public void print() {
        Node temp = start;
        StringBuilder s = new StringBuilder();
        while (temp != null) {
            s.append(temp.value).append(" <=> ");
            temp = temp.next;
        }
        System.out.println(s.substring(0, s.length() - 5));

    }

    public void push_front(int value) {
        Node t = new Node(value);
        if (start == null) {
            start = t;
            end = t;
            return;
        }
        t.next = start;
        start.prev = t;
        start = t;
    }

    public void insert(Node v, int value) {
        Node t = new Node(value);
        Node next = v.next;
        if (next == null) {
            t.next = null;
            v.next = t;
            t.prev = v;
            end = t;
            return;
        }
        next.prev = t;
        t.next = next;
        v.next = t;
        t.prev = v;
    }

    public void push_back(int value) {
        if (end == null) {
            Node t = new Node(value);
            start = t;
            end = t;
            return;
        }
        insert(end, value);
    }

    public void delete(Node v) {
        Node prev = v.prev;
        Node next = v.next;
        if (prev != null) {
            prev.next = next;
        }
        if (next != null) {
            next.prev = prev;
        }
        v.next = null;
        v.prev = null;
        if (v == start) {
            start = next;
        }
        if (v == end) {
            end = prev;
        }
    }

    public void pop_front() {
        delete(start);
    }

    public void pop_back() {
        delete(end);
    }

    public void revert() {
        if (start == null) {
            return;
        }
        Node a = start;
        Node b = start.next;
        a.next = null;
        if (b == null) {
            return;
        }
        Node c = b.next;
        while (b != null) {
            b.next = a;
            a.prev = b;
            a = b;
            b = c;
            if (c != null) {
                c = c.next;
            }
        }
        a.prev = null;
        Node t = start;
        start = end;
        end = t;
    }

    public void genList(int count) {
        int a = 0;
        int b = 100;
        int value = 0;
        for (int i = 0; i < count; i++) {
            value = (int) (Math.random() * ((b - a) + 1)) + a;
            push_back(value);
        }
    }

}
